eservice/pdo/sservice/block_store_manager.py

```diff
@@ -0,0 +1,285 @@
+            signing_hash_accumulator += b''.join(block_hashes)
+
+            signing_hash = hashlib.sha256(signing_hash_accumulator).digest()
+            signature = self.service_keys.sign(signing_hash, encoding=encoding)
+        except Exception as e :
+            logger.error("unknown exception packing response (BlockStatus); %s", str(e))
+            return StorageException('signature failed')
+
+        result = dict()
+        result['signature'] = signature
+        result['block_ids'] = list(map(encoding_fn, block_hashes))
+        return result
+
+    def check_block(self, block_id, encoding='b64') :
+        pass
+
+    def check_blocks(self, block_ids, encoding='b64') :
+        """Check status of a list of block
+
+        :param block_ids list of string: block identifiers
+        :param encoding string: encoding to use for block identifiers, raw/b64
+        :return list of dict: list of block status
+        """
+
+        decoding_fn = lambda x : x
+        if encoding == 'b64' :
+            decoding_fn = lambda x : base64.urlsafe_b64decode(x)
+
+        current_time = int(time.time())
+        mdb = self.block_store_env.open_db(b'meta_data')
+
+        block_status_list = []
+        with self.block_store_env.begin() as txn :
+            for block_id in block_ids :
+                # use the input format for the output block identifier
+                block_status = { 'block_id' : block_id, 'size' : 0, 'expiration' : 0 }
+                block_hash = decoding_fn(block_id)
+
+                raw_metadata = txn.get(block_hash, db=mdb)
+                if raw_metadata :
+                    metadata = BlockMetadata.unpack(raw_metadata)
+                    block_status['size'] = metadata.block_size
+                    block_status['expiration'] = metadata.expiration_time - current_time
+                    if block_status['expiration'] < 0 :
+                        block_status['expiration'] = 0
+
+                block_status_list.append(block_status)
+
+        return block_status_list
+
+    def expire_blocks(self) :
+        """Delete data and metadata for blocks that have expired
+        """
+        try :
+            mdb = self.block_store_env.open_db(b'meta_data')
+            bdb = self.block_store_env.open_db(b'block_data')
+
+            current_time = int(time.time())
+
+            count = 0
+            with self.block_store_env.begin() as txn :
+                cursor = txn.cursor(db=mdb)
+                for key, value in cursor :
+                    metadata = BlockMetadata.unpack(value)
+                    if metadata.expiration_time < current_time :
+                        logger.debug('expire block %s',base64.urlsafe_b64encode(key).decode())
+                        count += 1
+                        with self.block_store_env.begin(write=True) as dtxn :
+                            assert dtxn.delete(key, db=bdb)
```

Comment:
I guess, we might want to swap the order of these two deletes. Otherwise, if the combination : (Line 183 followed by a check_status) is executed by a different thread after the first delete, but before the second, we will simply update the meta-data and set expiration to be positive again, and the subsequent check_status will return a positive number all when the block is no longer present.

Reply:
there is already a higher level transaction that won't be committed. so the order shouldn't matter. that is, the two deletes must happen transactionally.